File: PY/Program/Store.py
from Album import Album
from Customer import Customer
from Employee import Employee
from Vip import VIPCustomer

class Store: # Store composite with Album, Customer, Employee, VIPCustomer
    def __init__(self): # Constructor
        self.albums = []
        self.customers = []
        self.employees = []
        self.vip_customers = []
        self._address = ""

    # Getter and Setter methods for address
    def set_address(self, address):
        self._address = address

    def get_address(self):
        return self._address

    # Methods to add data to the lists
    def add_album(self, album):
        self.albums.append(album)

    # Method to add customer
    def add_customer(self, customer):
        self.customers.append(customer)

    # Method to add employee
    def add_employee(self, employee):
        self.employees.append(employee)

    # Method to add VIP customer
    def add_vip_customer(self, vip):
        self.vip_customers.append(vip)

    # Method to print all data
    def print_all_data(self):
        print("\n=== Album List ===")
        if not self.albums: # Check if album list is empty because it will show different output
            print("Belum ada data album")
        else: # If not empty, print the album list
            print("--------------------------------------------------")
            print(f"{'Title':20}{'Group':15}{'Price':10}{'Stock':5}")
            print("--------------------------------------------------")
            for album in self.albums:
                album.print_album()
            print("--------------------------------------------------")

        print("\n=== Customer List ===")
        if not self.customers: # Check if customer list is empty because it will show different output
            print("Belum ada data customer")
        else: # If not empty, print the customer list
            print("--------------------------------------------------")
            print("ID\tPhone Number\tName\tAge\tFav. Group")
            print("--------------------------------------------------")
            for c in self.customers:
                c.print_customer()
            print("--------------------------------------------------")

        print("\n=== VIP Customer List ===")
        if not self.vip_customers: # Check if VIP customer list is empty because it will show different output
            print("Belum ada data customer VIP")
        else: # If not empty, print the VIP customer list
            print("-----------------------------------------------------------------------------------------------")
            print("ID\tPhone Number\tName \t  Age\t  Fav. Group\t Membership\tDiscount(%) Total Spent")
            print("-----------------------------------------------------------------------------------------------")
            for v in self.vip_customers:
                v.print_vip_customer()
            print("-----------------------------------------------------------------------------------------------")

        print("\n=== Employee List ===")
        if not self.employees: # Check if employee list is empty because it will show different output
            print("Belum ada data employee")
        else: # If not empty, print the employee list
            print("--------------------------------------------------------")
            print("ID\tName\tJenis Kelamin\tUmur\tPosition")
            print("--------------------------------------------------------")
            for e in self.employees:
                e.print_employee()
            print("--------------------------------------------------------")

File: CPP/Program/Store.cpp
#pragma once
#include <vector>
#include "Album.cpp"
#include "Customer.cpp"
#include "Employee.cpp"
#include "Vip.cpp"

// This class represents a Store that contains albums, customers, employees, and VIP customers
class Store {
    private:
        // Vectors to hold albums, customers, employees, and VIP customers
        vector<Album> albums;
        vector<Customer> customers;
        vector<Employee> employees;
        vector<VIPCustomer> vipCustomers;
        // Address of the store
        string address; 

    public:
        // Setters and Getters for address
        void setAddress(string address){
            this->address = address;
        }

        string getAddress(){
            return this->address;
        }
        
        // Methods to add albums
        void addAlbum(Album album) {
            albums.push_back(album);
        }
         // Methods to add customers
        void addCustomer(Customer customer) {
            customers.push_back(customer);
        }
        // Methods to add employees
        void addEmployee(Employee employee) {
            employees.push_back(employee);
        }
        // Methods to add VIP customers
        void addVIPCustomer(VIPCustomer vip) {
            vipCustomers.push_back(vip);
        }
        // Method to print all data in the store
        void printAllData() {
            cout << "\n=== Album List ===" << endl; // Print album list
            if(albums.empty()){ // Check if albums vector is empty or not to show different output
                cout << "Belum ada data album\n";
            }else{
                cout << "--------------------------------------------------" << endl;
                cout << left << setw(20) << "Title"
                    << setw(15) << "Group"
                    << setw(10) << "Price"
                    << setw(5)  << "Stock" << endl; // Print headers for album details
                cout << "--------------------------------------------------" << endl;
                for (int i = 0; i < albums.size(); i++) { // Loop through albums vector and print each album
                    albums[i].printAlbum();
                }
                cout << "--------------------------------------------------" << endl;
            }

            cout << "\n=== Customer List ===\n"; // Print customer list
            if(customers.empty()){ // Check if customers vector is empty or not to show different output
                cout << "Belum ada data customer\n";
            }else{
                cout << "--------------------------------------------------" << endl;
                cout << "ID\tPhone Number\tName\tAge\tFav. Group" << endl; // Print headers for customer details
                cout << "--------------------------------------------------" << endl;
                for (int i = 0; i < customers.size(); i++) { // Loop through customers vector and print each customer
                    customers[i].printCustomer();
                }
                cout << "--------------------------------------------------" << endl;
            }

            cout << "\n=== VIP Customer List ===\n"; // Print VIP customer list
            if(vipCustomers.empty()){ // Check if vipCustomers vector is empty or not to show different output
                cout << "Belum ada data customer VIP\n";
            }else{
                cout << "-----------------------------------------------------------------------------------------------" << endl;
                cout << "ID\tPhone Number\tName\t  Age\t  Fav. Group\t Membership\tDiscount(%) Total Spent" << endl; // Print headers for VIP customer details
                cout << "-----------------------------------------------------------------------------------------------" << endl;
                for (int i = 0; i < vipCustomers.size(); i++) { // Loop through vipCustomers vector and print each VIP customer
                    vipCustomers[i].printVIPCustomer();
                }
                cout << "-----------------------------------------------------------------------------------------------" << endl;
            }

            cout << "\n=== Employee List ===\n"; // Print employee list
            if(employees.empty()){ // Check if employees vector is empty or not to show different output
                cout << "Belum ada data employee\n"; 
            }else{
                cout << "--------------------------------------------------------" << endl;
                cout << "ID\tName\tJenis Kelamin\tUmur\tPosition" << endl; // Print headers for employee details
                cout << "--------------------------------------------------------" << endl;
                for (int i = 0; i < employees.size(); i++) { // Loop through employees vector and print each employee
                    employees[i].printEmployee();
                }
                cout << "--------------------------------------------------------" << endl;
            }
        }
};
